[Song Fang Yi] iP

songfangyl opened the pull request:
# NiHao
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

NiHao frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/songfangyl/ip/releases/tag/Level-9)
2. run in terminal with this command `java -jar duke.jar`
3. add your tasks.
4. let it manage your tasks for you :heart_eyes_cat:

And it is **FREE**!

Feature:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Reminders (coming soon)


If you Java programmer, you can use it to practice Java too. Here's the main method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

songfangyl committed:
Level 0,  Implement an initial skeletal version with greet and exit

songfangyl committed:
add interaction: chatbot now can echo what user input, exit with the keyword "exit"

songfangyl committed:
store feature

songfangyl committed:
list out the string when keyword "list" is scanned

songfangyl committed:
add Task class

songfangyl committed:
add mark and unmark features

songfangyl committed:
Debug for Level 3: change scanner.next()  to  scanner.nextLine()

songfangyl committed:
Add Three calsses: Deadline, Event, Todo and complete level 4 requirement

songfangyl committed:
change indentation of the output

songfangyl committed:
Automated Text UI Testing

songfangyl committed:
Error handling

songfangyl committed:
implement delete feature

songfangyl committed:
Update test case to test error handling and delete feature

songfangyl committed:
add javadoc

songfangyl committed:
Modify javadoc

songfangyl committed:
Add save/load feature

save the data in the data/duke.txt file.
the save format as below:
(isDone 1 or 0)/(description)/(by) or (from)/(to).

songfangyl committed:
Revert "Add save/load feature"

This reverts commit 235894b8b03bbfbe2e6172868f283502132637e9.

songfangyl committed:
Add save/load feature

save data into data/Duke.txt.
the format as below:
(isDone 1 or 0)/(description)/(by) or (from)/(to)

songfangyl committed:
Add the data/Duke.txt as a gitignore

songfangyl committed:
Merge branch 'branch-Level-7'

songfangyl committed:
First commit for branch level 8

songfangyl committed:
Update the date to java LocalDate

songfangyl committed:
Merge branch 'branch-Level-8'

songfangyl committed:
Change the design to OOP

Creating a lot of java files:
Parser: deals with making sense of the user command.

Storage: deals with loading tasks from the file and
saving tasks in the file.

Command: the action to be execute.

Ui: deals with interactions with the user.

TaskList: contains the task list e.g.,
it has operations to add/delete tasks in the list.

songfangyl committed:
Add gitignore to the save file

songfangyl committed:
Update DeleteCommand and EXPECTED.txt

songfangyl committed:
Merge branch 'branch-A-MoreOOP'

songfangyl committed:
Organize files by creating package

songfangyl committed:
Merge branch 'branch-A-Packages'

songfangyl committed:
Merge branch 'add-gradle-support' of https://github.com/songfangyl/ip

songfangyl committed:
Update build gradle

update the application mainClass.set, change it to the right path

songfangyl committed:
Create JUnit testcases

songfangyl committed:
Merge branch 'branch-A-JUnit'

songfangyl committed:
Debug Storage class

When building jar file, there is a problem where no data folder
in the directory.

songfangyl committed:
Implement find command

songfangyl committed:
Merge branch 'branch-Level-9'

songfangyl committed:
Javadoc

songfangyl committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java

songfangyl committed:
Implement equals method for each command

songfangyl committed:
Update JUnit test cases

songfangyl committed:
Correct coding style

songfangyl committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java

songfangyl committed:
Add checkstyle feature

songfangyl committed:
Merge branch 'branch-A-CheckStyle'

songfangyl committed:
Add GUI

Using fxml to build the gui,
also using css file to do the design which located at
resources/style.

songfangyl committed:
Update button and textfield design

songfangyl committed:
Merge branch 'branch-Level-10'

josepholim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,176 @@
+            if (detailArr.length <= 1) {
+                throw new DukeException("Ouuuu, I think you miss some information, try again!");
+            }
+            String des2 = detailArr[0];
```
Variable name can be more descriptive? Maybe `description` is better? 🤔

josepholim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,176 @@
+                throw new DukeException("Invalid mark command ?_? " +
+                        "this command should follow by only ONE INTEGER");
+            }
+            int i;
```
Variable name can be more descriptive? Maybe `index` is better? 🤔

josepholim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,176 @@
+                throw new DukeException("Invalid unmark command ?_? " +
+                        "this command should follow by only ONE INTEGER");
+            }
+            int j;
```
Try to be more consistent in variable naming, you used `i` for the mark case and `j` here

josepholim commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Constructor of Duke class.
```
Iirc the default style of javadoc recommends to add an empty line between the description and the parameter section 🤔

josepholim commented on src/main/java/duke/UiManager.java:
```diff
@@ -0,0 +1,120 @@
+package duke;
+import duke.task.Task;
```
Try to be consistent. You added an empty line between package and import in the TaskList.java file and other files, but not here :)

josepholim reviewed (commented):
LGTM 👍 :D

songfangyl committed:
Add assertions

Assertion define assumptions about the program state.

Many method in this project have some common assumption.

I add the assertion to majority of the class in UiManager and all
the command classes.

songfangyl committed:
Improve code quality

There is a very long method which is parse method in Parse class.
Refer to the SLAP, this method having multiple levels of
abstraction.

I extract the long method to several small methods.

By doing this this method succesfully follow the SLAP,
and the readability is increased.

Beside, I also change some of the variable names to be consistent,
when these variables inside the long method, there cannot be
more than one variable with the same name.

songfangyl committed:
Modify javadoc

There are several method have changed but the javadoc have not
changed yet.

I change all the javadoc into correct format.

songfangyl committed:
Delete ununsed code and make sure all the spacing is consistent

songfangyl committed:
Merge pull request #1 from songfangyl/branck-A-CodeQuality

Branch a code quality

songfangyl committed:
Merge branch 'master' into branch-A-Assertions

songfangyl committed:
Merge pull request #2 from songfangyl/branch-A-Assertions

Add assertions

songfangyl committed:
Add detect duplicate feature

There is no point to store multiple of same taks inside task list.

I add the detect duplicate feature to throw exception when user
try to add existing task into the task list.

I add a detectDuplicate method inside task list and invoke this
method when I add task into task list.

To check whether two task is same, I also override equals method in
Todo, Deadline and Event class.

songfangyl committed:
Merge branch 'branch-C-DetectDuplicates'

songfangyl committed:
Add welcome message when start the application

songfangyl committed:
Update javadoc and coding style

I update the javadoc to the correct format an add javadoc to those
method that missing javadoc.

For styles, I delete all the unuse method, unuse import.

songfangyl committed:
Update README.md

songfangyl committed:
Merge branch 'Branch-A-UserGuide'

songfangyl committed:
Update README.md

songfangyl committed:
Update Ui.png

songfangyl committed:
Update jar file name

songfangyl committed:
Update Junit Testcase

Update the Junit testcase to test the new command,
Update my Junit testcase to test for each possible testcase.

songfangyl committed:
Update README.md

songfangyl committed:
Update README.md
